### PR TITLE
auth context iterator

### PR DESCRIPTION
**Pull request by yang-g**

### CODE REVIEW

**Comment by yang-g** [comment]
Updated per the offline discussion.

File: include/grpc++/auth_context.h [hunk 1]
```diff
@@ -34,15 +34,42 @@
 #ifndef GRPCXX_AUTH_CONTEXT_H
 #define GRPCXX_AUTH_CONTEXT_H
 
+#include <iterator>
 #include <vector>
 
 #include <grpc++/config.h>
 
+struct grpc_auth_context;
```
Comment (jboeuf) [inline include/grpc++/auth_context.h]: Do you think we could put the iterator in another header? I think that would be cleaner. At least we would not "pollute" the AuthContext header (even though we would include the new iterator header).

File: include/grpc++/auth_context.h [hunk 2]
```diff
@@ -34,15 +34,42 @@
...
+    bool operator!=(const PropertyIterator& rhs) const;
+    const Property operator*();
+
+   private:
+    friend SecureAuthContext;
+    PropertyIterator();
+    PropertyIterator(const grpc_auth_property* property,
+                     const grpc_auth_property_iterator* iter);
+    const grpc_auth_property* property_;
+    // The following items form a grpc_auth_property_iterator.
```
Comment (jboeuf) [inline include/grpc++/auth_context.h]: It is unfortunate that you have to do that but I don't really see a way around it (except having a pointer to a `grpc_auth_property_iterator` but that would mean an extra alloc.

File: test/cpp/common/secure_auth_context_test.cc [hunk 3]
```diff
@@ -68,6 +69,51 @@ TEST_F(SecureAuthContextTest, Properties) {
...
+  EXPECT_EQ("chapi", p0.second);
+  EXPECT_EQ("name", p1.first);
+  EXPECT_EQ("chapo", p1.second);
+  EXPECT_EQ("foo", p2.first);
+  EXPECT_EQ("bar", p2.second);
+  ++iter;
+  EXPECT_EQ(context.end(), iter);
+  // Range-based for loop test.
+  int i = 0;
+  for (const AuthContext::Property p : context) {
```
Comment (jboeuf) [inline test/cpp/common/secure_auth_context_test.cc]: `const auto`?

File: include/grpc++/auth_context.h (hunk 1 above)
Comment (yang-g) [inline include/grpc++/auth_context.h]: Sure. The only lose is it is not under AuthContext any more. 

I guess I will add an AuthContextPropertyIterator in a public header (or under impl/?) and include it here. The main gain would be having its own .cc file to contain its implementation.

File: include/grpc++/auth_context.h (hunk 2 above)
Comment (yang-g) [inline include/grpc++/auth_context.h]: Exactly.

File: include/grpc++/auth_context.h (hunk 1 above)
Comment (jboeuf) [inline include/grpc++/auth_context.h]: I think that having another AuthPropertyIterator public header makes sense. This way the API remains kind of clean.

Comment (yang-g) [inline include/grpc++/auth_context.h]: Done.

File: test/cpp/common/secure_auth_context_test.cc (hunk 3 above)
Comment (yang-g) [inline test/cpp/common/secure_auth_context_test.cc]: Done.

**Comment by jboeuf** [comment]
Thanks much for the changes. LGTM.

**Pull request merged**